hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -767,8 +779,25 @@ public ClientConnection getRandomConnection() {
             }
         }
 
-        Iterator<TcpClientConnection> iterator = activeConnections.values().iterator();
-        return iterator.hasNext() ? iterator.next() : null;
+        Iterator<Map.Entry<UUID, TcpClientConnection>> iterator = activeConnections.entrySet().iterator();
+
+        Map.Entry<UUID, TcpClientConnection> connectionEntry = iterator.hasNext() ? iterator.next() : null;
+
+        if (connectionEntry != null) {
+            if (dataMember) {
+                UUID memberId = connectionEntry.getKey();
+
+                Member member = client.getClientClusterService().getMember(memberId);
+
+                if (member == null || member.isLiteMember()) {
+                    return null;
```

Comment:
This is a fallback code that's used when the load balancer returned `null`, even though it should return null only if there's no member available. But it's also a fallback when the load balancer doesn't implement the `nextDataMember` method, and in this case we should find the first data member. 

What if the first member returned by the iterator isn't a data member? A user upgrading to 4.2 that has a custom LB can get random "Client must be connected to at least one data member to execute SQL queries" errors.

Reply:
It's also used for non-smart client - can such client be connected to a lite member? Won't SQL work at all in that case?

Reply:
Not really. LB is only used when smart routing is enabled (see `isSmartRoutingEnabled` check). Otherwise, we always use the very first member connection (aka "unisocket" client). 

@sancar could you please give your comments on the correct behavior?

Reply:
Reworked the connection handling so that `dataMember=true` will return a connection to a data member, even if it is not the first one.